kroxylicious-api/src/main/java/io/kroxylicious/proxy/service/Contributor.java

```diff
@@ -7,46 +7,39 @@
 
 import io.kroxylicious.proxy.config.BaseConfig;
 
+import edu.umd.cs.findbugs.annotations.NonNull;
+
 /**
  * Support loading an Instance of a service, optionally providing it with configuration obtained
  * from the Kroxylicious configuration file.
  *
- * @param <T> the service type
+ * @param <S> the service type
+ * @param <C> the context type
  */
-public interface Contributor<T, S extends Context> {
+public interface Contributor<S, B extends BaseConfig, C extends Context<B>> {
 
     /**
-     * Identifies if this Contributor instance can create instances for the provided short name.
-     * @param typeName The type name to test for.
-     * @return <code>true</code> if the <code>typeName</code> is supported by this contributor.
+     * Identifies the type name this contributor offers.
+     * @return typeName
      */
-    boolean contributes(String typeName);
+    @NonNull
+    String getTypeName();
```

Comment:
I think everyone ( not sure about @SamBarker?) has supported the idea of switching to using classnames, but I've been trying to find things to defer to make this PR change less things.

Maybe it does make sense to include it in this PR though so it could all land as a single set of commits.

Reply:
I'm not against moving to classnames!

Reply:
Cool I'm in there making the change to support using the full Contributor classname. One thing I've hit on is if I allow the simple name then it's currently ambiguous because all the inner classes are called `Contributor`.

What do you think is the right thing?
1. Document that the Contributor class should have a unique name
2. Don't match on simple name for member classes, only compare simple name if it's a top level class

I think option 1 is easier to explain.

Reply:
Pushed up with option 2 for now

Reply:
another idea from our standup: the user isn't really interested in specifying which _Contributor_ they want to load. It might be better if they can specify the classname of the contributed thing. To do this we might need a method on Contributor like:

``` java
Class<? extends S> getServiceType();
```
and the framework would be responsible for matching the name (or simpleName) of that type.

Reply:
Should it be `Class<? extends S>` or `Class<S>`? I _think_ we want the latter, but I'll admit I'm struggling to explain to myself precisely why.

Reply:
Maybe the downside you're seeing is there is no guarantee that the type from `getServiceType` is the same as the type of the instance produced by `createInstance`, we could potentially enforce that with generics.

Reply:
Well, you can't guarantee that with Java anyway, because a subclass instance is, by definition, assignable to a subclass type. That means that it's possible for two different `Contributors` (with different `S`) to actually return the same instance type. But that's not unsafe or confusing. It just means we can't enforce a 1:1 relationship.

One thing the upper bound does allow is for `Contributor<X>` and `Contributor<Y>` (where `X` != `Y`) to both return the same `Z.class` (where Z is a subtype of X and Y). We've proposed a runtime check on all discoverable `Contributors` to disallow this because we want it to to be unambiguous which `Contributor` to use to instantiate the filter named in the config. So there's no harm from allowing the `? extends S`, other from the fact that it might be confusing (to whom?) if there are no `Contributor<Z>` on the classpath but the proxy allows a filter with `type: Z`.